### PR TITLE
[msg] Initialize session counters with random per spec.

### PR DESCRIPTION
**Pull request by turon**
#### Problem
Spec says to initialize session counters with random from [0, 2^28-1].

#### Change overview
Implement the specified behavior.

#### Testing
unit tests, CI, and manual observation on sniffer.

### CODE REVIEW

File: src/transport/MessageCounter.h [hunk 1]
```diff
@@ -106,8 +108,14 @@ class GlobalEncryptedMessageCounter : public MessageCounter
...
 public:
-    static constexpr uint32_t kInitialValue = 1;
-    LocalSessionMessageCounter() : value(kInitialValue) {}
+    static constexpr uint32_t kInitialValue = 1;    ///< Used for initializing peer counter
+
+    /**
+     * Initialize a local message counter with random value between [0, 2^28-1].
+     */
+    LocalSessionMessageCounter() {
+        value = Crypto::GetRandU32() & 0x0FFFFFFF;
```
Comment (emargolis) [inline src/transport/MessageCounter.h:117]: Maybe introduce and use here new constant:
`constexpr suint32_t k28BitMask = 0x0FFFFFFF;`

**Review by emargolis — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

File: src/transport/MessageCounter.h [hunk 2]
```diff
@@ -106,8 +107,12 @@ class GlobalEncryptedMessageCounter : public MessageCounter
 class LocalSessionMessageCounter : public MessageCounter
 {
 public:
-    static constexpr uint32_t kInitialValue = 1;
-    LocalSessionMessageCounter() : value(kInitialValue) {}
+    static constexpr uint32_t kInitialValue = 1; ///< Used for initializing peer counter
+
+    /**
+     * Initialize a local message counter with random value between [0, 2^28-1].
```
Comment (electrocucaracha) [inline src/transport/MessageCounter.h:113]: Are you referring to [this spec](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/secure_channel/Message_Format.adoc#111-message-counter-initialization)? If that's the case it should be nice to include some context.
```suggestion
     * Initialize a local message counter with random value between [0, 2^28-1]. This increases the difficulty of traffic analysis attacks by making it harder to determine how long a particular session has been open.
```

**Review by woody-apple — APPROVED** [review]

File: src/transport/MessageCounter.h (hunk 1 above)
Comment (turon) [reply src/transport/MessageCounter.h]: I applied the suggestion, but in general constexpr increases flash size relative to a #define macro or enum.

File: src/transport/MessageCounter.h [hunk 3]
```diff
@@ -106,8 +107,14 @@ class GlobalEncryptedMessageCounter : public MessageCounter
 class LocalSessionMessageCounter : public MessageCounter
 {
 public:
-    static constexpr uint32_t kInitialValue = 1;
-    LocalSessionMessageCounter() : value(kInitialValue) {}
```
Comment (Damian-Nordic) [inline src/transport/MessageCounter.h:110]: Did you remove the constructor intentionally? Is the value initialized to `kInitialValue` somewhere else?

**Review by Damian-Nordic — APPROVED** [review]

**Pull request merged**